### ISSUE TITLE
Type narrowing after assignment does not work for class variables

### ISSUE THREAD

**Issue by last-partizan** [microsoft/pyright#3662]
Consider following case:

**Example code**
```python
from typing import Optional


class ApiViewBase:
    lookup_key: Optional[str] = None


class ApiView(ApiViewBase):
    lookup_key = "test"

    def get_object(self) -> str:
        objects: dict[str, str] = {}
        reveal_type(self.lookup_key)
        return objects[self.lookup_key]


key: Optional[str] = None
key = "is narrowed to literal"

reveal_type(key)
```

**Actual behavior (pyright)**
```
/tmp/test.py
  /tmp/test.py:13:21 - information: Type of "self.lookup_key" is "str | None"
  /tmp/test.py:14:16 - error: Argument of type "str | None" cannot be assigned to parameter "__k" of type "str" in function "__getitem__"
    Type "str | None" cannot be assigned to type "str"
      Type "None" cannot be assigned to type "str" (reportGeneralTypeIssues)
  /tmp/test.py:20:13 - information: Type of "key" is "Literal['is narrowed to literal']"
1 error, 0 warnings, 2 informations
```

**Expected behavior**
I expect in both cases type to be narrowed to str (or literal). mypy handles this correctly, narrowing to str:

```
mypy /tmp/test.py
/tmp/test.py:13: note: Revealed type is "builtins.str"
/tmp/test.py:20: note: Revealed type is "builtins.str"
Success: no issues found in 1 source file
```

**VS Code extension or command-line**
- pyright 1.1.257
